site/guides/gas-sponsorship-eligibility.md

```diff
@@ -0,0 +1,67 @@
+---
+
+# How to Check if a User Operation is Eligible for Sponsorship
+
+As mentioned from the previous guide on [How to Sponsor Gas for a User Operation](./sponsoring-gas.md), with Account Kit dponsor gas fees for transactions via the [Gas Manager](https://docs.alchemy.com/docs/gas-manager-services).
+
+But what happens when the user operation you are sending fails to satisfy the criteria set in the gas manager policy? If you do send the user operation that is **not** eligible for the gas sponsorship under your Gas Manager policy, the user operation will fail and thrown an error. Then how would you know a user operation's eligibility for gas sponsorship in advance?
+
+This guide explains how to check if a user operation is eligibile for sponsor gas before sending the user operation, so that you can handle the user experience based on the gas sponsorship eligibility accordingly.
+
+## How to Check for Gas Sponsorship Eligibility for a User Operation
+
+First, you can follow the same instructions from the previous guide on [How to Sponsor Gas for a User Operation](./sponsoring-gas.md) to set up your `SmartAccountProvider` and link your gas policy. Then, instead of calling `sendUserOperation` on the provider right away, you can use [`buildUserOperation`](../packages/aa-core/provider/buildUserOperation.md) to check if the user operation is eligible for gas sponsorship or not.
+
+Calling `buildUserOperation` method on the provider that is configured with the gas manager (`withAlchemyGasManager`) will run the middleware pipeline including the paymaster middleware, which will thrown an error if the user operation is not eligible for gas sponsorship under the policy.
+
+::: code-group
+
+```ts [check-gas-sponsorship-eligibility.ts]
+import { provider } from "./provider.ts";
+
+// Find your Gas Manager policy id at: // [!code focus:10]
+//dashboard.alchemy.com/gas-manager/policy/create
+const GAS_MANAGER_POLICY_ID = "YourGasManagerPolicyId";
+
+// Link the provider with the Gas Manager. This ensures user operations
+// sent with this provider get sponsorship from the Gas Manager.
+provider.withAlchemyGasManager({
+  policyId: GAS_MANAGER_POLICY_ID,
+});
+
+// If not eligible for gas sponsorship, this will thrown an error
+const uoStruct = await provider
+  .buildUserOperation({
+    target: "0xTargetAddress",
+    data: "0xCallData",
+    value: 0n, // value in bigint or leave undefined
+  })
+  .catch((err) => {
+    // if the user operation is not eligible for gas sponsorship, error will be thrown and caught here
+    console.error(err);
+  });
+```
+
+<<< @/snippets/provider.ts
+
+:::
+
+We recommend you use this method to check for the gas sponsorship eligibility prior to sending the user operation. This way, you can handle the different cases of gas sponsonship eligibility accordingly for your users.
```

Comment:
maybe we should recommend that this be handled within the `withPaymasterAndData` middleware override instead? ie. if the RPC call errors they can choose to fallback to 0x. 

We should probably also update the `withAlchemyGasManager` method to support a `fallbackToWallet` param that will do this handling for devs

Reply:
> maybe we should recommend that this be handled within the withPaymasterAndData middleware override instead? ie. if the RPC call errors they can choose to fallback to 0x.

Will make the update.

> We should probably also update the withAlchemyGasManager method to support a fallbackToWallet param that will do this handling for devs

As discussed in the morning sync, will set up a time to more formal scoping/arch design. https://app.asana.com/0/1205598840815267/1205892709620241/f

Reply:
assuming we'll chat about this design next week? should i hold off on reviewing this?

Reply:
yea let's update this doc once we're aligned on approach. I like the what you suggested @denniswon of having override in `sendUserOperation` instead so that this works for all paymaster middleware. So instead of `fallbackToWallet` being part of the `withAlchemyGasManager` we can make it part of the `UserOperationOverrides` object as `paymaster.fallbackToAccount`

Reply:
https://www.notion.so/alchemotion/Eng-Spec-Flexible-Gas-Manager-Policy-Handling-ef69e8f7cafa41839f34cd2e9906c2c1?pvs=4 @avasisht23